File: codes/java/684.冗余连接.java

import java.util.HashMap;

class UnionFind {
	private HashMap<Integer, Integer> parent;

	UnionFind() {
		parent = new HashMap<>();
	}

	public Integer find(Integer i) {
		while (parent.get(i) != null) {
			i = parent.get(i);
		}
		return i;
	}

	public void union(Integer p, Integer q) {
		Integer pRoot = find(p);
		Integer qRoot = find(q);
		if (pRoot == qRoot) {
			return;
		}
		parent.put(qRoot, pRoot);
	}
}

class Solution {
	public int[] findRedundantConnection(int[][] edges) {
		UnionFind uf = new UnionFind();
		for (int[] edg : edges) {
			int parent = edg[0];
			int children = edg[1];
			if (uf.find(parent) == uf.find(children)) {
				return edg;
			}
			uf.union(parent, children);
		}
		return new int[0];
	}
}

File: codes/golang/684.冗余连接.go

type UnionFind struct {
	parent map[int]int
}

func NewUnionFind() *UnionFind {
	return &UnionFind{parent: make(map[int]int)}
}

func (uf *UnionFind) find(i int) int {
	if _, ok := uf.parent[i]; !ok {
		uf.parent[i] = i
		return i
	}
	if uf.parent[i] != i {
		uf.parent[i] = uf.find(uf.parent[i])
	}
	return uf.parent[i]
}

func (uf *UnionFind) union(p, q int) {
	parent := uf.find(p)
	children := uf.find(q)
	if parent == children {
		return
	}
	uf.parent[children] = parent
}

func findRedundantConnection(edges [][]int) []int {
	uf := NewUnionFind()
	for _, edg := range edges {
		parent := edg[0]
		children := edg[1]
		if uf.find(parent) == uf.find(children) {
			return edg
		}
		uf.union(parent, children)
	}
	return []int{}
}
